src/util/server.ml

```diff
@@ -0,0 +1,155 @@
+open Batteries
+open Jsonrpc
+
+type t = {
+  file: Cil.file;
+  do_analyze: Analyses.increment_data -> Cil.file -> unit;
+}
+
+module type Command = sig
+  val name: string
+
+  type args
+  type result
+
+  val args_of_yojson: Yojson.Safe.t -> args Ppx_deriving_yojson_runtime.error_or
+  val result_to_yojson: result -> Yojson.Safe.t
+
+  val process: args -> t -> result
+end
+
+module Registry = struct
+  type t = (string, (module Command)) Hashtbl.t
+  let make () : t = Hashtbl.create 32
+  let register (reg: t) (module R : Command) = Hashtbl.add reg R.name (module R)
+end
+
+let registry = Registry.make ()
+
+let handle_exn id exn =
+  Response.Error.(make Code.InternalError (Printexc.to_string exn) () |> Response.error id)
+
+module ParamParser (C : Command) = struct
+  let parse params =
+    let args = 
+      params
+      |> Option.map_default Message.Structured.to_json `Null
+      |> C.args_of_yojson
+    in
+    match args with
+    | Ok args -> Ok args
+    | Error err ->
+      (* This is a hack to handle cases where C.args is primitive type like int or string. *)
+      match params with
+      | Some `List [param] -> (
+          match C.args_of_yojson param with
+          | Ok arg -> Ok arg
+          | _ -> Error err)
+      | _ -> Error err
```

Comment:
What exactly is a problem with those?

Also, I think it might be better to not allow them anyway, but have them be in records, where the field name (key) is at least descriptive (like the reset field for the analyze command below), rather than just having values like `5` with no apparent meaning.

Reply:
jsonrpc expects `params` to be an `` `Assoc `` or `` `List ``, and this fails for simple types like `string` and `int`. So, as a workaround, if the initial parsing fails, I check if the `params` is a list that contains a single object, and I try to parse this object.

I guess I could remove it. Originally, I used it for `analyze`, but I also thought a single boolean argument looked ugly and replaced it with a record type.

Reply:
Thanks for the explanation, I wasn't too familiar with what what JSON-RPC allows there.
I guess we can leave this fallback as well, might be a bit more convenient for experimenting, although all commands that we intend to keep using should use records.